src/main/java/Duke.java

```diff
@@ -1,10 +1,50 @@
 public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+
+    public Duke(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (DukeException e) {
+            ui.showLoadingError();
+            tasks = new TaskList();
+        }
+    }
+
+    public void run() {
+        ui.welcomeMessage();
+        tasks.showCurrentTasks();
+        ui.showLineBreak();
+        boolean isFinished = false;
+        while (!(isFinished)) {
+            try {
+            String inputFromUser = ui.handleInput();
+               Command c = Parser.parse(inputFromUser);
+            if (c == null) {
+                throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-( ☹ OOPS!!!");
+            }
+               tasks.runCommand(c);
+            if (c.getCommand().equals("bye")) {
+                isFinished = true;
+            }
+        } catch (DukeException e) {
+                System.out.println("Invalid command received: " + e.getMessage());
```

Comment:
You have many System.out.println that are not done by UI. You might want to have user interactions done solely by UI instead.

Reply:
Yes, but it would take more efforts on our side, since you can see we need to have something in the current context e.g. `e.getMessage()` to be printed out. Hence it would take more efforts to put it in the class `Ui`. However, I think this is a good suggestion and we should put every println in class `Ui` to avoid having "side effects" in our `Main`/`Duke` class.